db/kv_channel_storage.go

```diff
@@ -30,29 +30,33 @@ const (
 	kSequenceOffsetLength = 0 // disabled until we actually need it
 )
 
-// ChannelStorage implemented as an interface, to support swapping to different underlying storage model
+// ChannelStorage implemented as two interfaces, to support swapping to different underlying storage model
 // without significant refactoring.
-type ChannelStorage interface {
+type ChannelStorageReader interface {
+	// GetAllEntries returns all entries for the channel in the specified range, for all vbuckets
+	GetChanges(fromSeq base.SequenceClock, channelClock base.SequenceClock, limit int) ([]*LogEntry, error)
+}
+
+type ChannelStorageWriter interface {
 
 	// AddEntrySet adds a set of entries to the channel index
 	AddEntrySet(entries []*LogEntry) (clockUpdates base.SequenceClock, err error)
 
-	// GetAllEntries returns all entries for the channel in the specified range, for all vbuckets
-	GetChanges(fromSeq base.SequenceClock, channelClock base.SequenceClock) ([]*LogEntry, error)
-
 	// If channel storage implementation uses separate storage for log entries and channel presence,
 	// WriteLogEntry and ReadLogEntry can be used to read/write.  Useful when changeIndex wants to
 	// manage these document outside the scope of a channel.  StoresLogEntries() allows callers to
 	// check whether this is available.
 	StoresLogEntries() bool
 	WriteLogEntry(entry *LogEntry) error
+}
 
-	// For unit testing only
-	getIndexBlockForEntry(entry *LogEntry) IndexBlock
+type ChannelStorage interface {
+	ChannelStorageReader
+	ChannelStorageWriter
 }
 
-func NewChannelStorage(bucket base.Bucket, channelName string, partitions *base.IndexPartitions) ChannelStorage {
-	return NewBitFlagStorage(bucket, channelName, partitions)
+func NewChannelStorageReader(bucket base.Bucket, channelName string, partitions *base.IndexPartitions) ChannelStorageReader {
+	return NewDenseStorageReader(bucket, channelName, partitions)
```

Comment:
I've had conversation with several gophers that it's better to return the impl rather than the interface.  And recently I converted some code in sg-bucket and it avoided some type assertions.

What about changing to:

```
unc NewChannelStorageReader(bucket base.Bucket, channelName string, partitions *base.IndexPartitions) *DenseStorageReader {
```

instead?  Will that break anything?

I can try to dig up more behind the rationale if needed